### ISSUE TITLE
Puma today - general thoughts & UNIXSockets, PR #1851

### ISSUE THREAD

**Issue by MSP-Greg** [puma/puma#1878]
There have been several issues that seem to pertain to restarts, and many involve UNIXSockets.  Some seem to date to 3.12.0, some later, and the later ones may not specify whether the issue has existed in the past.

A thought that has gone through my mind in the past few weeks is that shutdown code seems to be unaware of whether some form of restart/respawn is is also happening, and whether it overlaps with the shutdown (phased restart, graceful shutdown, respawn a worker, etc).

Issue #1775 shows the code for Binder#close at the time.  UNIXSockets are unlinked in a block, but a failure on one will stop the iteration.  It was then updated at PR #1784 (26-Apr), which allowed the iteration to complete on error.  At the time, there were no `Errno::ECONNREFUSED` errors appearing in the Travis CI logs.

The `Errno::ECONNREFUSED` errors began appearing in #1846 'Test cleanup and parallelization', and can be seen in [Travis 3555](https://travis-ci.org/puma/puma/builds/559705658), but not [Travis 3552](https://travis-ci.org/puma/puma/builds/559698463).

The `unlink` code was then removed by #1829 and merged on 09-July. `Errno::ECONNREFUSED` errors are still raised.  A small change to a test was included, but it simply checks for the existence of the file, which may not show whether the UNIXSockets reconnected successfully.

So, PR #1851 was an attempt to return to using `unlink`, but with tighter criteria for which files that is done to.  The tests went green.

Summing up, I think more tests need to written.  We're trying to fix issues without tests, and long term, things will be inadvertently broken again.

I'm comfortable writing tests for ssl issues; I can test them locally, and my knowledge of SSL is largely cross-platform.  Without tests for issues involving `fork` and `UNIXSockets`, all I can do is rely on the tests.  I think we're all tired of issues reappearing...

**Comment by MSP-Greg** [comment]
Above was kind of a rant about 'more tests'.  Closing...

@dentarg Thanks.

**Comment by nateberkopec** [comment]
That's valuable information if we can turn this into a "todo" item that says "these specific LOC are undertested in ways XYZ"

**Comment by MSP-Greg** [comment]
Everything I addressed above has been changed by merged or current PR's.  I think the issues are best addressed there...
